core/src/main/java/com/graphhopper/routing/util/GenericFlagEncoder.java

```diff
@@ -0,0 +1,478 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for 
+ *  additional information regarding copyright ownership.
+ * 
+ *  GraphHopper GmbH licenses this file to you under the Apache License, 
+ *  Version 2.0 (the "License"); you may not use this file except in 
+ *  compliance with the License. You may obtain a copy of the License at
+ * 
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.util;
+
+import com.graphhopper.reader.OSMRelation;
+import com.graphhopper.reader.OSMWay;
+import com.graphhopper.util.EdgeIteratorState;
+import java.util.*;
+import java.util.Map.Entry;
+
+/**
+ * This encoder tries to store all way information into a 64bit value. Later extendable to multiple
+ * ints or bytes. The assumption is that edge.getFlags is cheap and can be later replaced by e.g.
+ * one or more (cheap) calls of edge.getData(index).
+ * <p>
+ * Currently limited to motor vehicle but later could handle different modes like foot or bike too.
+ *
+ * @author Peter Karich
+ */
+public class GenericFlagEncoder extends AbstractFlagEncoder
+{
+    private long bit0;
+    private EncodedDoubleValue carFwdMaxspeedEncoder;
+    private EncodedDoubleValue carBwdMaxspeedEncoder;
+    private EncodedValue surfaceEncoder;
+    private final Map<String, Integer> surfaceMap = new HashMap<>();
+    private EncodedValue highwayEncoder;
+    private final Map<String, Integer> highwayMap = new HashMap<>();
+    private EncodedValue transportModeEncoder;
+    private final List<String> transportModeList = new ArrayList<>();
+    private final Map<String, Integer> transportModeMap = new HashMap<>();
+
+    public GenericFlagEncoder()
+    {
+        // TODO include turn information
+        super(5, 5, 0);
+
+        maxPossibleSpeed = 140;
+
+        //
+        // TODO do something about the access values (forestry, agricultural, destination, private, delivery, customers)
+        //
+        List<String> highwayList = Arrays.asList(
+                /* reserve index=0 for unset roads (not accessible) */
+                "_default",
+                "motorway", "motorway_link", "motorroad",
+                "trunk", "trunk_link",
+                "primary", "primary_link", "secondary", "secondary_link", "tertiary", "tertiary_link",
+                "unclassified", "residential", "living_street", "service", "road", "track",
+                "ferry", "shuttle");
```

Comment:
> There is no need to be sorry :)

:) sure :+1: !

Yes, this is used in the route tag but I was trying to avoid a new tag key here as there are no combinations of e.g. motorway and ferry or with shuttle. This is just to avoid using yet another set of bits.